File: backend/app/routers/yields.py
from datetime import datetime, timedelta, timezone
from enum import Enum
from typing import Optional

from fastapi import APIRouter, Depends, HTTPException, Query
from sqlalchemy import and_, or_, func, text
from sqlalchemy.orm import Session, joinedload

from app.config.stablecoins import STABLECOIN_SYMBOLS
from app.dependencies import get_db
from app.models.yield_opportunity import YieldOpportunity, YieldSnapshot
from app.schemas import YieldOpportunityListOut, YieldOpportunityDetailOut, YieldHistoryPoint, ProtocolOut

router = APIRouter()


class SortOrder(str, Enum):
    APY_DESC = "apy_desc"
    APY_ASC = "apy_asc"
    TVL_DESC = "tvl_desc"
    TVL_ASC = "tvl_asc"


PERIOD_DAYS = {"7d": 7, "30d": 30, "90d": 90}


@router.get("/yields", response_model=dict)
def get_yields(
    category: Optional[str] = Query(None),
    sort: SortOrder = Query(SortOrder.APY_DESC),
    tokens: Optional[str] = Query(None),
    vault_tag: Optional[str] = Query(None),
    stablecoins_only: bool = Query(False),
    limit: int = Query(100, ge=1, le=500),
    offset: int = Query(0, ge=0),
    db: Session = Depends(get_db),
):
    q = db.query(YieldOpportunity).filter(YieldOpportunity.is_active == True)  # noqa: E712

    if category:
        q = q.filter(YieldOpportunity.category == category)

    if vault_tag:
        q = q.filter(YieldOpportunity.extra_data["vault_tag"].astext == vault_tag)

    if tokens:
        token_list = [t.strip() for t in tokens.split(",")]
        q = q.filter(YieldOpportunity.tokens.overlap(token_list))

    if stablecoins_only:
        q = q.filter(YieldOpportunity.apy_current > 0)
        q = q.filter(
            or_(
                # Multiply: only stable/rwa loops (excludes JLP, SOL, BONKSOL directional pairs)
                and_(
                    YieldOpportunity.category == "multiply",
                    YieldOpportunity.extra_data["vault_tag"].astext.in_(
                        ["stable_loop", "rwa_loop"]
                    ),
                ),
                # Non-multiply (lending, vault, etc.): at least one stablecoin token
                and_(
                    YieldOpportunity.category != "multiply",
                    YieldOpportunity.tokens.overlap(list(STABLECOIN_SYMBOLS)),
                ),
                # PT-* tokens (Exponent principal tokens — fixed yield, stablecoin exposure)
                func.exists(
                    text(
                        "SELECT 1 FROM unnest(yield_opportunities.tokens) AS t"
                        " WHERE t LIKE 'PT-%'"
                    )
                ),
            )
        )

    if sort == SortOrder.APY_DESC:
        q = q.order_by(YieldOpportunity.apy_current.desc().nullslast())
    elif sort == SortOrder.APY_ASC:
        q = q.order_by(YieldOpportunity.apy_current.asc().nullsfirst())
    elif sort == SortOrder.TVL_DESC:
        q = q.order_by(YieldOpportunity.tvl_usd.desc().nullslast())
    elif sort == SortOrder.TVL_ASC:
        q = q.order_by(YieldOpportunity.tvl_usd.asc().nullsfirst())

    count_col = func.count().over().label("_total")
    rows = q.add_columns(count_col).offset(offset).limit(limit).all()
    total = rows[0][1] if rows else 0
    results = [row[0] for row in rows]
    last_updated = max((r.updated_at for r in results if r.updated_at), default=None)

    items = []
    for r in results:
        item = YieldOpportunityListOut.model_validate(r)
        if r.extra_data:
            item.protocol_url = r.extra_data.get("protocol_url")
        items.append(item)

    return {
        "data": items,
        "meta": {"total": total, "last_updated": last_updated, "limit": limit, "offset": offset},
    }


@router.get("/yields/{yield_id}", response_model=YieldOpportunityDetailOut)
def get_yield_detail(
    yield_id: int,
    db: Session = Depends(get_db),
):
    opp = db.query(YieldOpportunity).options(joinedload(YieldOpportunity.protocol)).filter(YieldOpportunity.id == yield_id).first()
    if not opp:
        raise HTTPException(status_code=404, detail="Yield opportunity not found")

    since = datetime.now(timezone.utc) - timedelta(days=7)
    snapshots = (
        db.query(YieldSnapshot)
        .filter(
            YieldSnapshot.opportunity_id == yield_id,
            YieldSnapshot.snapshot_at >= since,
        )
        .order_by(YieldSnapshot.snapshot_at.asc())
        .all()
    )

    item = YieldOpportunityListOut.model_validate(opp)
    if opp.extra_data:
        item.protocol_url = opp.extra_data.get("protocol_url")
    data = item.model_dump()
    data["extra_data"] = opp.extra_data
    data["deposit_address"] = opp.deposit_address
    data["protocol"] = ProtocolOut.model_validate(opp.protocol) if opp.protocol else None
    data["recent_snapshots"] = [YieldHistoryPoint.model_validate(s) for s in snapshots]

    return YieldOpportunityDetailOut(**data)


@router.get("/yields/{yield_id}/history", response_model=dict)
def get_yield_history(
    yield_id: int,
    period: str = Query("7d"),
    limit: int = Query(500, ge=1, le=2000),
    offset: int = Query(0, ge=0),
    db: Session = Depends(get_db),
):
    if not db.query(YieldOpportunity).filter(YieldOpportunity.id == yield_id).first():
        raise HTTPException(status_code=404, detail="Yield opportunity not found")

    days = PERIOD_DAYS.get(period, 7)
    since = datetime.now(timezone.utc) - timedelta(days=days)

    q = (
        db.query(YieldSnapshot)
        .filter(
            YieldSnapshot.opportunity_id == yield_id,
            YieldSnapshot.snapshot_at >= since,
        )
        .order_by(YieldSnapshot.snapshot_at.asc())
    )

    total = q.count()
    snapshots = q.offset(offset).limit(limit).all()

    return {
        "data": [YieldHistoryPoint.model_validate(s) for s in snapshots],
        "meta": {"total": total, "limit": limit, "offset": offset},
    }

File: backend-ts/src/discover/routes/yields.ts
import type { FastifyInstance } from "fastify";
import {
  sql,
  eq,
  and,
  or,
  asc,
  gte,
  arrayOverlaps,
} from "drizzle-orm";
import { db } from "../db/connection.js";
import { yieldOpportunities, yieldSnapshots, protocols } from "../db/schema.js";
import { NotFoundError } from "../../shared/error-handler.js";
import { STABLECOIN_SYMBOLS } from "../../shared/constants.js";
import { YieldsQuery, YieldHistoryQuery } from "./schemas.js";

const PERIOD_DAYS: Record<string, number> = { "7d": 7, "30d": 30, "90d": 90 };

function numOrNull(val: string | null | undefined): number | null {
  if (val === null || val === undefined) return null;
  const n = Number(val);
  return Number.isFinite(n) ? n : null;
}

export async function yieldsRoutes(app: FastifyInstance) {
  // -----------------------------------------------------------------------
  // GET /yields
  // -----------------------------------------------------------------------
  app.get("/yields", {
    config: { rateLimit: { max: 60, timeWindow: "1 minute" } },
    handler: async (request) => {
      const q = YieldsQuery.parse(request.query);

      const conditions: ReturnType<typeof eq>[] = [
        eq(yieldOpportunities.is_active, true),
      ];

      if (q.category) {
        conditions.push(eq(yieldOpportunities.category, q.category));
      }

      if (q.vault_tag) {
        conditions.push(
          sql`${yieldOpportunities.extra_data}->>'vault_tag' = ${q.vault_tag}`,
        );
      }

      if (q.tokens) {
        const tokenList = q.tokens.split(",").map((t) => t.trim());
        conditions.push(arrayOverlaps(yieldOpportunities.tokens, tokenList));
      }

      if (q.stablecoins_only) {
        conditions.push(
          sql`${yieldOpportunities.apy_current} > 0`,
        );
        const stableArr = [...STABLECOIN_SYMBOLS];
        conditions.push(
          or(
            // Multiply: stable_loop or rwa_loop
            and(
              eq(yieldOpportunities.category, "multiply"),
              sql`${yieldOpportunities.extra_data}->>'vault_tag' IN ('stable_loop', 'rwa_loop')`,
            ),
            // Non-multiply: at least one stablecoin token
            and(
              sql`${yieldOpportunities.category} != 'multiply'`,
              arrayOverlaps(yieldOpportunities.tokens, stableArr),
            ),
            // PT-* tokens
            sql`EXISTS (SELECT 1 FROM unnest(${yieldOpportunities.tokens}) AS t WHERE t LIKE 'PT-%')`,
          )!,
        );
      }

      // Sort
      let orderBy;
      switch (q.sort) {
        case "apy_asc":
          orderBy = sql`${yieldOpportunities.apy_current} ASC NULLS FIRST`;
          break;
        case "tvl_desc":
          orderBy = sql`${yieldOpportunities.tvl_usd} DESC NULLS LAST`;
          break;
        case "tvl_asc":
          orderBy = sql`${yieldOpportunities.tvl_usd} ASC NULLS FIRST`;
          break;
        default: // apy_desc
          orderBy = sql`${yieldOpportunities.apy_current} DESC NULLS LAST`;
      }

      const rows = await db
        .select({
          opp: yieldOpportunities,
          _total: sql<number>`count(*) over()`.as("_total"),
        })
        .from(yieldOpportunities)
        .where(and(...conditions))
        .orderBy(orderBy)
        .offset(q.offset)
        .limit(q.limit);

      const total = rows.length > 0 ? Number(rows[0]._total) : 0;

      let lastUpdated: Date | null = null;
      const data = rows.map((r) => {
        const o = r.opp;
        if (o.updated_at && (!lastUpdated || o.updated_at > lastUpdated)) {
          lastUpdated = o.updated_at;
        }
        const extra = o.extra_data as Record<string, unknown> | null;
        return {
          id: o.id,
          protocol_id: o.protocol_id,
          external_id: o.external_id,
          name: o.name,
          category: o.category,
          tokens: o.tokens,
          apy_current: numOrNull(o.apy_current),
          apy_7d_avg: numOrNull(o.apy_7d_avg),
          apy_30d_avg: numOrNull(o.apy_30d_avg),
          tvl_usd: numOrNull(o.tvl_usd),
          min_deposit: numOrNull(o.min_deposit),
          lock_period_days: o.lock_period_days ?? 0,
          risk_tier: o.risk_tier,
          protocol_name: o.protocol_name,
          is_active: o.is_active,
          max_leverage: numOrNull(o.max_leverage),
          utilization_pct: numOrNull(o.utilization_pct),
          liquidity_available_usd: numOrNull(o.liquidity_available_usd),
          is_automated: o.is_automated,
          depeg: numOrNull(o.depeg),
          protocol_url: extra?.protocol_url ?? null,
          updated_at: o.updated_at,
        };
      });

      return {
        data,
        meta: {
          total,
          last_updated: lastUpdated,
          limit: q.limit,
          offset: q.offset,
        },
      };
    },
  });

  // -----------------------------------------------------------------------
  // GET /yields/:id
  // -----------------------------------------------------------------------
  app.get<{ Params: { id: string } }>("/yields/:id", {
    config: { rateLimit: { max: 60, timeWindow: "1 minute" } },
    handler: async (request) => {
      const yieldId = Number(request.params.id);

      const oppRows = await db
        .select({
          opp: yieldOpportunities,
          protocol: protocols,
        })
        .from(yieldOpportunities)
        .leftJoin(protocols, eq(yieldOpportunities.protocol_id, protocols.id))
        .where(eq(yieldOpportunities.id, yieldId))
        .limit(1);

      if (oppRows.length === 0) {
        throw new NotFoundError("Yield opportunity not found");
      }

      const { opp, protocol } = oppRows[0];

      const since = new Date(Date.now() - 7 * 24 * 60 * 60 * 1000);
      const snapshots = await db
        .select()
        .from(yieldSnapshots)
        .where(
          and(
            eq(yieldSnapshots.opportunity_id, yieldId),
            gte(yieldSnapshots.snapshot_at, since),
          ),
        )
        .orderBy(asc(yieldSnapshots.snapshot_at));

      const extra = opp.extra_data as Record<string, unknown> | null;

      return {
        id: opp.id,
        protocol_id: opp.protocol_id,
        external_id: opp.external_id,
        name: opp.name,
        category: opp.category,
        tokens: opp.tokens,
        apy_current: numOrNull(opp.apy_current),
        apy_7d_avg: numOrNull(opp.apy_7d_avg),
        apy_30d_avg: numOrNull(opp.apy_30d_avg),
        tvl_usd: numOrNull(opp.tvl_usd),
        min_deposit: numOrNull(opp.min_deposit),
        lock_period_days: opp.lock_period_days ?? 0,
        risk_tier: opp.risk_tier,
        protocol_name: opp.protocol_name,
        is_active: opp.is_active,
        max_leverage: numOrNull(opp.max_leverage),
        utilization_pct: numOrNull(opp.utilization_pct),
        liquidity_available_usd: numOrNull(opp.liquidity_available_usd),
        is_automated: opp.is_automated,
        depeg: numOrNull(opp.depeg),
        protocol_url: extra?.protocol_url ?? null,
        updated_at: opp.updated_at,
        extra_data: opp.extra_data,
        deposit_address: opp.deposit_address,
        protocol: protocol
          ? {
              id: protocol.id,
              slug: protocol.slug,
              name: protocol.name,
              description: protocol.description,
              website_url: protocol.website_url,
              logo_url: protocol.logo_url,
              audit_status: protocol.audit_status,
              auditors: protocol.auditors,
              integration: protocol.integration,
            }
          : null,
        recent_snapshots: snapshots.map((s) => ({
          snapshot_at: s.snapshot_at,
          apy: numOrNull(s.apy),
          tvl_usd: numOrNull(s.tvl_usd),
        })),
      };
    },
  });

  // -----------------------------------------------------------------------
  // GET /yields/:id/history
  // -----------------------------------------------------------------------
  app.get<{ Params: { id: string } }>("/yields/:id/history", {
    config: { rateLimit: { max: 60, timeWindow: "1 minute" } },
    handler: async (request) => {
      const yieldId = Number(request.params.id);
      const q = YieldHistoryQuery.parse(request.query);

      // Verify opportunity exists
      const exists = await db
        .select({ id: yieldOpportunities.id })
        .from(yieldOpportunities)
        .where(eq(yieldOpportunities.id, yieldId))
        .limit(1);

      if (exists.length === 0) {
        throw new NotFoundError("Yield opportunity not found");
      }

      const days = PERIOD_DAYS[q.period] ?? 7;
      const since = new Date(Date.now() - days * 24 * 60 * 60 * 1000);

      const whereClause = and(
        eq(yieldSnapshots.opportunity_id, yieldId),
        gte(yieldSnapshots.snapshot_at, since),
      );

      const [countResult, snapshots] = await Promise.all([
        db
          .select({ count: sql<number>`count(*)` })
          .from(yieldSnapshots)
          .where(whereClause),
        db
          .select()
          .from(yieldSnapshots)
          .where(whereClause)
          .orderBy(asc(yieldSnapshots.snapshot_at))
          .offset(q.offset)
          .limit(q.limit),
      ]);

      return {
        data: snapshots.map((s) => ({
          snapshot_at: s.snapshot_at,
          apy: numOrNull(s.apy),
          tvl_usd: numOrNull(s.tvl_usd),
        })),
        meta: {
          total: Number(countResult[0].count),
          limit: q.limit,
          offset: q.offset,
        },
      };
    },
  });
}
